src/rgw/rgw_lua_utils.cc

```diff
@@ -67,10 +66,91 @@ void open_standard_libs(lua_State* L) {
   unsetglobal(L, "dofile");
   unsetglobal(L, "debug");
   // remove os.exit()
-  lua_getglobal(L, "os");
-  lua_pushstring(L, "exit");
-  lua_pushnil(L);
-  lua_settable(L, -3);
+  if (lua_getglobal(L, "os") == LUA_TTABLE) {
+    lua_pushstring(L, "exit");
+    lua_pushnil(L);
+    lua_settable(L, -3);
+  }
+}
+
+// allocator function that verifies against maximum allowed memory value
+void* allocator(void* ud, void* ptr, size_t osize, size_t nsize) {
+  auto mem = reinterpret_cast<int*>(ud); // remaining memory
+  // free memory
+  if (nsize == 0) {
+    if (mem && *mem > 0 && ptr) {
```

Comment:
I don't see a reason for these `*mem > 0` tests here and elsewhere.

Reply:
That's interesting. Is that a defensive programming measure or did you encounter an allocation issue in your testing?

Reply:
when running the unit tests, there are always allocations done after the call to `lua_close()`.

Reply:
Well, I'm not sure how serious this hardening effort is to prevent malicious actors but you can easily work around your memory limit protection using something like:

```
local function a(str) return a(str..str) end
t = setmetatable({}, {__gc = function() a('.') end})
```

Which will exhaust memory as soon as you run lua_close.

Reply:
> Well, I'm not sure how serious this hardening effort is to prevent malicious actors but you can easily work around your memory limit protection using something like:
> 
> ```
> local function a(str) return a(str..str) end
> t = setmetatable({}, {__gc = function() a('.') end})
> ```
> 
> Which will exhaust memory as soon as you run lua_close.

the main goal is not malicious actors, but regular users that do not understand the memory implications fo their scripts.
regardless, maybe a better option is to give some grace memory during lua_close, just to make sure that the cleanup could be done even if the main script used all of the memory.

Reply:
I suspect that if a script runs out of memory then any allocation failures by finalizers called during `lua_close` are irrelevant.

Reply:
@batrick i made the changes according to the comments, can you please re-review?

Reply:
The change looks good to me but does it not make the `*mem > 0` now irrelevant?